edpop_explorer/marc21.py

```diff
@@ -0,0 +1,218 @@
+    _place_field_subfield = ('264', 'a')
+    _dating_field_subfield = ('264', 'c')
+    _extent_field_subfield = ('300', 'a')
+    _physical_description_field_subfield = ('300', 'b')
+    _size_field_subfield = ('300', 'c')
+    _fingerprint_field_subfield = ('026', 'e')
+
+    @classmethod
+    @abstractmethod
+    def _get_link(cls, data: Marc21Data) -> Optional[str]:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def _get_identifier(cls, data: Marc21Data) -> Optional[str]:
+        pass
+
+    @classmethod
+    def _marc21data_to_record(cls, data: Marc21Data) -> Marc21BibliographicalRecord:
+        record = Marc21BibliographicalRecord(from_reader=cls)
+        record.data = data
+        record.link = cls._get_link(data)
+        record.identifier = cls._get_identifier(data)
+        # NOTE: it is probably better to outfactor the following logic to
+        # other class methods, to offer more flexibility and because this
+        # will become more complex as we add normalization.
+        title = data.get_first_subfield(*cls._title_field_subfield)
+        if title:
+            record.title = Field(title)
+        alternative_title = data.get_first_subfield(
+            *cls._alternative_title_field_subfield
+        )
+        if alternative_title:
+            record.alternative_title = Field(alternative_title)
+        publisher = data.get_first_subfield(*cls._publisher_field_subfield)
+        if publisher:
+            record.publisher_or_printer = Field(publisher)
+        place = data.get_first_subfield(*cls._place_field_subfield)
+        if place:
+            record.place_of_publication = Field(place)
+        language = data.get_first_subfield(*cls._language_field_subfield)
+        # TODO: look up if this field is repeatable - if so support multiple
+        # languages
+        if language:
+            language_field = LanguageField(language)
+            language_field.normalize()
+            record.languages = [language_field]
+        dating = data.get_first_subfield(*cls._dating_field_subfield)
+        if dating:
+            record.dating = Field(dating)
+        extent = data.get_first_subfield(*cls._extent_field_subfield)
+        if extent:
+            record.extent = Field(extent)
+        physical_description = data.get_first_subfield(
+            *cls._physical_description_field_subfield
+        )
+        if physical_description:
+            record.physical_description = Field(physical_description)
+        size = data.get_first_subfield(*cls._size_field_subfield)
+        if size:
+            record.size = Field(size)
+        fingerprint = data.get_first_subfield(*cls._fingerprint_field_subfield)
+        if fingerprint:
+            record.fingerprint = Field(fingerprint)
+
+        # Add the contributors
+        record.contributors = cls._get_contributors(data)
+
+        # Add the holdings
+        record.holdings = cls._get_holdings(data)
+
+        return record
```

Comment:
This function is very long and also rather repetitive. Please outfactor the repeated pattern so you can make this shorter and easier to read.

Reply:
I agree 100% here, but this code already existed (in a module that had a different name before) and the problem is addressed in #21. The best solution would probably be to create one method for every field that is extracted, but that requires a large deal of refactoring throughout the readers.